### ISSUE TITLE
scalar formatting for custom types (literal vs folded)

### ISSUE THREAD

**Issue by fholzer** [nodeca/js-yaml#427]
I tried implementing a custom type for Ansible's vault type for use of encrypted variables in variable files. (In contrast to completely encrypted files.) My target is to be able to read Ansible variable files, and write them. I'm not interested in de/encrypting any data, I just need to serialize the vault type such that it is equivalent to the input file.

For testing I created a repository containing the type I created, an example input file (input.yml) and a test script that just reads the input file and writes the parsed data to output.yml.
The problem is that I couldn't find any mechanisms to
* control how the data I return in the `represent` function of my type is serialized in the ouput file. Meaning whether the string scalar i return is serialized as literal, or folded. In the input the string is a literal, in the ouput it is folded.
* control how the type name is encoded. In the input it just says `!vault`, while in the output it says `!<!vault>`

When serializing, ja-yaml Ideally a `diff` between input.yml and output.yml should say that there is no difference at all between the two files.

* Official documentation: https://docs.ansible.com/ansible/2.4/vault.html#use-encrypt-string-to-create-encrypted-variables-to-embed-in-yaml
* My test repository: https://github.com/fholzer/js-yaml-ansible-test

**Comment by jthomerson** [comment]
This is probably essentially a duplicate of #59 and #66 
It would be nice to have the option to dump literal style instead of folded style, for the same reasons @fholzer mentions: comparison of input vs output values with no-diff.

**Comment by rlidwka** [comment]
> control how the type name is encoded. In the input it just says !vault, while in the output it says !<!vault>

Fixed in js-yaml 4.0.0.

> control how the data I return in the represent function of my type is serialized in the ouput file. Meaning whether the string scalar i return is serialized as literal, or folded. In the input the string is a literal, in the ouput it is folded.

Here's a dirty hack:

If you define schema as implicit, its represent function will return result as is (no transformations applied to it whatsoever)

proof of concept (it will generate invalid yaml if you are not careful):

```js
const yaml = require('js-yaml')


function Format(data) {
  if (!(this instanceof Format)) return new Format(data);
  this.data = data;
}


let CustomFormatType = new yaml.Type('!format', {
  kind: 'scalar',
  resolve: () => false,
  instanceOf: Format,
  represent: f => f.data
});


let schema = yaml.DEFAULT_SCHEMA.extend({ implicit: [ CustomFormatType ] });

function replacer(key, value) {
  if (value === 2) return Format(`'any\ncode:\nwhatsoever"`)
  return value;
}

console.log(yaml.dump({ foo: 1, bar: 2 }, { schema, replacer }));
/* outputs:

foo: 1
bar: 'any
code:
whatsoever"
*/
```
